src/timesignature.ts

```diff
@@ -0,0 +1,142 @@
+    );
+  }
+
+  numbers.forEach((number) => {
+    if (isNaN(Number(number))) {
+      throw new Vex.RERR('BadTimeSignature', `Invalid time spec: ${timeSpec}. Must contain two valid numbers.`);
+    }
+  });
+};
+
+export class TimeSignature extends StaveModifier {
+  point?: number;
+  bottomLine?: number;
+  protected timeSig?: TimeSignatureInfo;
+
+  protected validate_args: boolean;
+
+  topLine?: number;
+
+  static get CATEGORY(): string {
+    return 'timesignatures';
+  }
+
+  static get glyphs(): Record<string, TimeSignatureGlyphInfo> {
+    return {
+      C: {
+        code: 'timeSigCommon',
+        point: 40,
+        line: 2,
+      },
+      'C|': {
+        code: 'timeSigCutCommon',
+        point: 40,
+        line: 2,
+      },
+    };
+  }
+
+  constructor(timeSpec?: string, customPadding = 15, validate_args = true) {
+    super();
+    this.setAttribute('type', 'TimeSignature');
+    this.validate_args = validate_args;
+
+    if (!timeSpec) return;
+    const padding = customPadding;
+
+    this.point = this.musicFont.lookupMetric('digits.point');
+    const fontLineShift = this.musicFont.lookupMetric('digits.shiftLine', 0);
+    this.topLine = 2 + fontLineShift;
+    this.bottomLine = 4 + fontLineShift;
+    this.setPosition(StaveModifier.Position.BEGIN);
+    this.timeSig = this.parseTimeSpec(timeSpec);
+    this.setWidth(this.getTimeSig().glyph.getMetrics().width!);
+    this.setPadding(padding);
+  }
+
+  getCategory(): string {
+    return TimeSignature.CATEGORY;
+  }
+
+  parseTimeSpec(timeSpec: string): TimeSignatureInfo {
+    if (timeSpec === 'C' || timeSpec === 'C|') {
+      const { line, code, point } = TimeSignature.glyphs[timeSpec];
+      return {
+        line,
+        num: false,
+        glyph: new Glyph(code, point),
+      };
+    }
+
+    if (this.validate_args) {
+      assertIsValidFraction(timeSpec);
+    }
+
+    const [topDigits, botDigits] = timeSpec.split('/').map((number) => number.split(''));
+
+    return {
+      num: true,
+      glyph: this.makeTimeSignatureGlyph(topDigits, botDigits),
+    };
+  }
+
+  makeTimeSignatureGlyph(topDigits: string[], botDigits: string[]): Glyph {
+    const glyph = new TimeSignatureGlyph(this, topDigits, botDigits, 'timeSig0', this.point!);
+    return glyph;
+  }
+
+  getTimeSig(): TimeSignatureInfo {
+    return this.timeSig!;
+  }
+
+  setTimeSig(timeSpec: string): this {
+    this.timeSig = this.parseTimeSpec(timeSpec);
+    return this;
+  }
+
+  draw(): void {
+    if (!this.x) {
+      throw new Vex.RERR('TimeSignatureError', "Can't draw time signature without x.");
+    }
+
+    if (!this.stave) {
+      throw new Vex.RERR('TimeSignatureError', "Can't draw time signature without stave.");
+    }
+
+    this.setRendered();
+    this.timeSig!.glyph.setStave(this.stave);
+    this.timeSig!.glyph.setContext(this.stave.getContext());
+    this.placeGlyphOnLine(this.timeSig!.glyph, this.stave, this.timeSig!.line!);
```

Comment:
Why is there an "!" after .line"!"?   For numeric time signatures,  this.timeSig does not have a line.

Reply:
You are right! This was also the case in the original code. Do you know where this line should come from?

Reply:
@0xfe @ronyeh this example shows that disabling `@typescript-eslint/no-non-null-assertion` is not a good idea. I agree with @ronyeh we should enable it again and use `// eslint-disable-next-line` with a lot of care.

Reply:
I am not a fan of using the non-null assertion (!) operator for two reasons:

1) Its presence in the code suggests that the code could be improved in the future through better type information. We are using TypeScript, so we should strive to benefit from type checking.

2) It's kind of difficult to spot while reading code, and it's also difficult to search for via CTRL/CMD+F, because the ! is also the not operator. While `// eslint-disable-next-line` doesn't look as elegant, it's a big red flag to future readers of the code. When we revisit those lines post TS migration, we could explore ways to refactor to avoid the `// eslint-disable-next-line`.

I am also not a fan of over-using optional properties (? operator) for a similar reason. When I was migrating the Parser, I realized that the Results interface has a lot of optional properties. https://github.com/ronyeh/vexflow/blob/8318a4113f3439af1e23fd23b61f049a89231e80/src/parser.ts#L36  Some of the properties are only used in the tokenizer/lexer stage, and other properties are only used in the parser stage. In the future, it might be nice to break it up into two: LexerResult vs. ParserResult. This might clean up some of the code later on, where we have to check for the presence of properties before proceeding.

Reply:
I generally agree with all of this, and it's good practice to be explicit about things, have sane defaults, minimize optional properties etc. But we're migrating a relatively large decade-old code base with lots of warts (JS itself evolved significantly over that period). This is already tiresome, so I'm hesitant to put up too many barriers and increase friction.

I've also noticed that without `!`, there's a tendency to use things like `?? 0` or `|| default_value`, which IMO is a lot worse because they lead to silent hard-to-debug incorrect behavior. I'd rather have the code fail early and loud with a clear stack trace than the alternative, and lean on heavy unit testing to force as many code paths as possible.

My suggestion would be to minimize the use of things like `any`, `!`, `?` wherever possible. If the pure approach incurs too much gymnastics, then use a shortcut (like `any` or `!`), and deal with it later (add a TODO, file an issue, etc.) We're still gaining a ton from the other stuff.

Reply:
@0xfe Do you know where this line should come from?

Reply:
This is effectively a y-shift, so it can default to `0`. You can make this non-optional.

Reply:
Done